abide/src/main/scala/scala/tools/abide/traversal/SimpleWarnings.scala

```diff
@@ -0,0 +1,162 @@
+package scala.tools.abide.traversal
+
+import scala.reflect.macros._
+import scala.language.experimental.macros
+import scala.tools.abide.traversal._
+
+/**
+ * IncrementalWarnings
+ *
+ * Base trait that provides an extension to [[RuleState]] which requires a transformation
+ * that adds a new warning to the accumulated set:
+ * ```scala
+ * def nok(warning: Warning): State
+ * ```
+ *
+ * We also refine the [[State]] type member to require an [[IncrementalState]] to
+ * guarantee all subtypes rules can register warnings with the `nok` transformer.
+ */
+trait IncrementalWarnings extends TraversalRule {
+  import context.universe._
+
+  /**
+   * IncrementalState
+   *
+   * Extension to [[RuleState]] that requires a transformer to register warnings
+   */
+  trait IncrementalState extends RuleState {
+    def nok(warning: Warning): State
+  }
+
+  /**
+   * The state type we'll be dealing with in subtype rules, required to be a subtype
+   * of the [[IncrementalState]] trait.
+   *
+   * @see [[IncrementalState]]
+   */
+  type State <: IncrementalState
+
+  /** Register a warning */
+  def nok(warning: Warning): Unit = { transform(_ nok warning) }
```

Comment:
Yes, that sounds like a good idea. States probably shouldn't offer their transformers as API.